### PR TITLE
Allow and preserve image width and height

### LINKED ISSUE #14201: Allow and preserve image width and height attributes
After checking the PoC check in #14147, we decided that we want to go further with this approach.

### Scope
* load the `width` and `height` attributes from HTML (done in the PoC).
* breaking change: model `width`-> `resizedWidth`, `widthAttribute` -> `width` (it was not a part of the PoC).
* breaking change: streamlined the `srcset` attribute in the model. On master it is `{ data: "...", width: "" }` but since now width has been delegated to I`mageSizeAttributes`, I figured this object form is useless (done in the PoC).
* resize stays as is but we add `height: auto` to the styles to preserve the proper aspect ratio. (Is it a BC?)

We need to test it against our current upload adapters if this doesn't break anything.

### PR DESCRIPTION
**Pull request by mmotyczynska**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Feature (image): It should be possible to load and preserve image `width` and `height` attributes. Closes #14201.

MAJOR BREAKING CHANGE (image): The name of the image resize `width` attribute has been changed to `resizedWidth`. The `width` and `height` attributes are now used to preserve image width and height.

MAJOR BREAKING CHANGE (image): The `srcset` attribute has been simplified. It is no longer an object `{ data: "...", width: "..." }`, but a value previously stored in the `data` part.

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

File: packages/ckeditor5-image/tests/image/imageblockediting.js [hunk 1]
```diff
@@ -651,7 +572,7 @@ describe( 'ImageBlockEditing', () => {
 					'<imageBlock ' +
 						'src="/assets/sample.png" ' +
 						'alt="alt text" ' +
-						'srcset=\'{ "data": "small.png 148w, big.png 1024w", "width": "1024" }\'>' +
+						'srcset=\'{"data":"small.png 148w, big.png 1024w"}\'>' +
```
Comment (arkflpc) [inline packages/ckeditor5-image/tests/image/imageblockediting.js:575]: Why is it not: `'srcset="small.png 148w, big.png 1024w">'`?

File: packages/ckeditor5-image/tests/image/imageediting.js [hunk 2]
```diff
@@ -235,7 +193,7 @@ describe( 'ImageEditing', () => {
 					'<imageBlock ' +
 						'src="/assets/sample.png" ' +
 						'alt="alt text" ' +
-						'srcset=\'{ "data": "small.png 148w, big.png 1024w", "width": "1024" }\'>' +
+						'srcset=\'{"data":"small.png 148w, big.png 1024w"}\'>' +
```
Comment (arkflpc) [inline packages/ckeditor5-image/tests/image/imageediting.js:196]: Unwrap from `"data"`.

File: packages/ckeditor5-image/tests/image/imageediting.js [hunk 3]
```diff
@@ -255,19 +213,19 @@ describe( 'ImageEditing', () => {
 					'<paragraph><imageInline ' +
 						'src="/assets/sample.png" ' +
 						'alt="alt text" ' +
-						'srcset=\'{ "data": "small.png 148w, big.png 1024w", "width": "1024" }\'>' +
+						'srcset=\'{"data":"small.png 148w, big.png 1024w"}\'>' +
```
Comment (arkflpc) [inline packages/ckeditor5-image/tests/image/imageediting.js:216]: Unwrap from `"data"`.

File: packages/ckeditor5-image/tests/image/imageediting.js [hunk 4]
```diff
@@ -1027,7 +879,7 @@ describe( 'ImageEditing', () => {
 					'<imageBlock ' +
 						'src="/assets/sample.png" ' +
 						'alt="alt text" ' +
-						'srcset=\'{ "data": "small.png 148w, big.png 1024w", "width": "1024" }\'>' +
+						'srcset=\'{"data":"small.png 148w, big.png 1024w"}\'>' +
```
Comment (arkflpc) [inline packages/ckeditor5-image/tests/image/imageediting.js:882]: Unwrap from `"data"`.

File: packages/ckeditor5-image/tests/image/imageediting.js [hunk 5]
```diff
@@ -1047,7 +899,7 @@ describe( 'ImageEditing', () => {
 					'<paragraph><imageInline ' +
 						'src="/assets/sample.png" ' +
 						'alt="alt text" ' +
-						'srcset=\'{ "data": "small.png 148w, big.png 1024w", "width": "1024" }\'>' +
+						'srcset=\'{"data":"small.png 148w, big.png 1024w"}\'>' +
```
Comment (arkflpc) [inline packages/ckeditor5-image/tests/image/imageediting.js:902]: Unwrap from `"data"`.

File: packages/ckeditor5-image/tests/image/imageblockediting.js [hunk 6]
```diff
@@ -184,7 +162,7 @@ describe( 'ImageBlockEditing', () => {
...
```
Comment (arkflpc) [inline packages/ckeditor5-image/tests/image/imageblockediting.js:165]: Unwrap from `"data"`.

File: packages/ckeditor5-image/tests/image/imageinlineediting.js [hunk 7]
```diff
@@ -190,19 +168,19 @@ describe( 'ImageInlineEditing', () => {
 					'<paragraph><imageInline ' +
 						'src="/assets/sample.png" ' +
 						'alt="alt text" ' +
-						'srcset=\'{ "data": "small.png 148w, big.png 1024w", "width": "1024" }\'>' +
+						'srcset=\'{"data":"small.png 148w, big.png 1024w"}\'>' +
```
Comment (arkflpc) [inline packages/ckeditor5-image/tests/image/imageinlineediting.js:171]: Unwrap from `"data"`.

File: packages/ckeditor5-image/tests/image/imageinlineediting.js [hunk 8]
```diff
@@ -680,7 +603,7 @@ describe( 'ImageInlineEditing', () => {
...
```
Comment (arkflpc) [inline packages/ckeditor5-image/tests/image/imageinlineediting.js:606]: Unwrap from `"data"`.

**Review by arkflpc — CHANGES_REQUESTED** [review]

File: packages/ckeditor5-image/tests/image/imageblockediting.js (hunk 1 above)
Comment (mmotyczynska) [reply packages/ckeditor5-image/tests/image/imageblockediting.js]: Updated in b389967.

File: packages/ckeditor5-image/tests/image/imageediting.js (hunk 2 above)
Comment (mmotyczynska) [reply packages/ckeditor5-image/tests/image/imageediting.js]: Updated in b389967.

File: packages/ckeditor5-image/tests/image/imageediting.js (hunk 3 above)
Comment (mmotyczynska) [reply packages/ckeditor5-image/tests/image/imageediting.js]: Updated in b389967.

File: packages/ckeditor5-image/tests/image/imageediting.js (hunk 4 above)
Comment (mmotyczynska) [reply packages/ckeditor5-image/tests/image/imageediting.js]: Updated in b389967.

File: packages/ckeditor5-image/tests/image/imageediting.js (hunk 5 above)
Comment (mmotyczynska) [reply packages/ckeditor5-image/tests/image/imageediting.js]: Updated in b389967.

File: packages/ckeditor5-image/tests/image/imageblockediting.js (hunk 6 above)
Comment (mmotyczynska) [reply packages/ckeditor5-image/tests/image/imageblockediting.js]: Updated in b389967.

File: packages/ckeditor5-image/tests/image/imageinlineediting.js (hunk 7 above)
Comment (mmotyczynska) [reply packages/ckeditor5-image/tests/image/imageinlineediting.js]: Updated in b389967.

File: packages/ckeditor5-image/tests/image/imageinlineediting.js (hunk 8 above)
Comment (mmotyczynska) [reply packages/ckeditor5-image/tests/image/imageinlineediting.js]: Updated in b389967.

**Review by arkflpc — APPROVED** [review]

**Comment by arkflpc** [comment]
@ckeditor/qa-team , can you look at it?
@mmotyczynska , will you provide more info if neede?

**Comment by dufipl** [comment]
@arkflpc Sure, we will take a look at this 🕵️

**Comment by mmotyczynska** [comment]
The first  change:

*   We now enable `width` and `height` attributes for `imageInline` and `imageBlock` elements.
*   Unfortunately the `width` attribute in model element was already in use for image resize. So we had to change its name from `width` to `resizedWidth`.

The second change:

*   We simplified the `srcset` attribute which previosusly was an object: `{ data: "...", width: "..." }`, but now it's a value previously stored in the `data` part. This is because we handle the `width` in the first change and it's no longer needed here in `srcset`.

Manual tests:

*   There is one new [manual test](http://localhost:8125/ckeditor5-image/tests/manual/imagesizeattributes.html)  that already loads new attributes for block/inline images with and without resize value.
*   But these changes may affect also other tests that handle images, e.g. paste from office.

What to check:

*   The image should load and preserve new attributes (use dedicated manual test or `setData()` for example).
*   Images should preserve aspect ratio.
*   This change should not break the image resize feature, but its value should now be stored in `resizedWidth` instead of `width` attribute.
*   Other than that use your creativity and try to break the editor or break importing images etc. :smile:

**Comment by Acrophost** [comment]
## `height` attribute is completely ignored even when `width` is not specified

### Steps
1. Open the [manual test](http://localhost:8125/ckeditor5-image/tests/manual/imagesizeattributes.html)
2. Call `editor.setData('Foo');` in the console

### Result
Image gets the default size of 800x376 regardless of given height.
The height attribute is kept but size itself is overwritten by `.ck-content .image-inline picture, .ck-content .image-inline img { height: auto; } `.
If the width is specified, image resizes accordingly to it.

**Comment by Acrophost** [comment]
## [Safari][regression] Pasting images from Word results in wall of errors
### Steps
1. Open [manual test](http://localhost:8125/ckeditor5-image/tests/manual/imagesizeattributes.html) on Safari
2. Copy a word document containing images
3. Paste the document into the editor

### Result
Images do not paste.
Console throws some errors.
Example document: [example.docx](https://github.com/ckeditor/ckeditor5/files/11597593/example.docx)
Stack trace:
```javascript
[Error] Refused to connect to blob:http://localhost:8125/e12ecd8e-92f2-4926-ab1f-19ff732943d1 because it does not appear in the connect-src directive of the Content Security Policy.

[Error] Refused to connect to blob:http://localhost:8125/55a55881-3e29-4ce1-9534-32cbfa5501d7 because it does not appear in the connect-src directive of the Content Security Policy.

[Error] Refused to connect to blob:http://localhost:8125/8da4bc75-144b-481b-bbc0-51c95a8e05a2 because it does not appear in the connect-src directive of the Content Security Policy.

[Error] Refused to load blob:http://localhost:8125/e12ecd8e-92f2-4926-ab1f-19ff732943d1 because it does not appear in the img-src directive of the Content Security Policy.

[Error] Refused to load blob:http://localhost:8125/55a55881-3e29-4ce1-9534-32cbfa5501d7 because it does not appear in the img-src directive of the Content Security Policy.

[Error] Refused to load blob:http://localhost:8125/8da4bc75-144b-481b-bbc0-51c95a8e05a2 because it does not appear in the img-src directive of the Content Security Policy.
```

**Comment by Acrophost** [comment]
We have finished testing ✨ 
The main thing we noticed is that height is getting overwritten by `height: auto` style and thus, has no effect even when there is no width defined (when this declaration is disabled, all works as it should).
Aside from that, everything looks good to us 🎉 

@charlttsie @dufipl thank you for the jolly cooperation 🍻

**Pull request merged**